File: EC530-Multi-threads-Extra-Credits/EC530-multi-threads.py
import random
import threading
import time

# initialize global variable and lock
count = 0
lock = threading.Lock()


# thread worker, sleep for random seconds between 1 - 5
def thread_func(id):
    global count
    lock.acquire()
    print(f"Thread {id}: count = {count}")
    time.sleep(random.randint(1, 5))
    count += 1
    print(f"Thread {id}: count++")
    lock.release()


if __name__ == '__main__':
    thread_num = 4
    threads = []
    thread_ids = []

    for i in range(thread_num):
        thread_ids.append(i)

    # create threads and run
    for i in range(thread_num):
        t = threading.Thread(target=thread_func, args=(thread_ids[i],))
        threads.append(t)
        t.start()

    # wait for threads to complete
    for t in threads:
        t.join()

    print("Threads completed")

File: EC530-Multi-threads-Extra-Credits/EC530_multi-threads.cpp
//
// Created by Renyu Jiang on 3/21/23.
//
#include <cstdio>
#include <cstdlib>
#include <pthread.h>
#include <unistd.h>

// initialize global variable and mutex
int count = 0;
pthread_mutex_t mutex;

// thread worker, sleep for random seconds between 1 - 5
void *thread_func(void *arg) {
    pthread_mutex_lock(&mutex);
    printf("Thread %d: count = %d\n", *(int *)arg, count);
    sleep(rand() % 5 + 1);
    count++;
    printf("Thread %d: count++\n", *(int *)arg);
    pthread_mutex_unlock(&mutex);
    return nullptr;
}

int main() {
    int thread_num=4;
    pthread_t threads[thread_num];
    int thread_ids[thread_num]; // thread id

    for(int i=0;i<thread_num;i++){
        thread_ids[i]=i;
    }

    // initialize mutex
    pthread_mutex_init(&mutex, nullptr);

    // create threads and run
    for (int i = 0; i < thread_num; i++) {
        if (pthread_create(&threads[i], nullptr, thread_func, &thread_ids[i]) != 0) {
            exit(-1);
        }
    }

    // wait for threads to complete
    for (int i = 0; i < thread_num; i++) {
        if (pthread_join(threads[i], nullptr) != 0) {
            exit(-1);
        }
    }

    printf("Threads completed\n");

    // destroy mutex
    pthread_mutex_destroy(&mutex);

    return 0;
}
